Enable GPU execution of atm_compute_moist_coefficients via OpenACC

abishekg7 opened the pull request:
This PR enables the GPU execution of atm_compute_moist_coefficients subroutine.

Tested with a real and idealized test cases.

gdicker1 commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -1730,39 +1730,60 @@ subroutine atm_compute_moist_coefficients( dims, state, diag, mesh, &
+
+
+      !$acc parallel
+      !$acc loop
```
I think the loop directives should be more explicit (e.g. `loop gang worker` here). Just to match with other work we're doing.

gdicker1 commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -1730,39 +1730,60 @@ subroutine atm_compute_moist_coefficients( dims, state, diag, mesh, &
          end if
       end do
+      !$acc end parallel
+      !$acc exit data copyout(cqw, cqu, qtot) delete(scalars)
```
This `exit data` directive should be surrounded by `MPAS_ACC_TIMER_{START,STOP}` calls.

mgduda replied:
As I commented on another PR, I'd also advocate for at least splitting the `copyout` and `delete` clauses onto separate lines, or better yet, to turn them into separate directives. I think this will make it easier when it comes to shifting around and fusing data directives later, and at least for me, it just makes it easier to visually recognize that there are two things happening in this `exit data` directive.

gdicker1 reviewed (changes requested):
This may need some additional work. When I compared baseline and test runs of the regional testcase, I get differences in two fields in the history file. Restart files are reported as matching.

Output from compare_netcdf.py:
```
Comparing files: f1 - f2                                                                                                                             
        f1=../2024Dec31_MPASAPR1223_OpenACC/baseline_acc_1735844488/history.2019-09-01_00.06.00.nc                                                   
        f2=test_att1735694999/history.2019-09-01_00.06.00.nc                                                                                         

            Variable  Min       Max       
=========================================
initial_time is not a numeric field and will not be compared
xtime is not a numeric field and will not be compared
        depv_dt_fric -32.448421  33.315609
     depv_dt_fric_pv -32.448421  30.517973
mminlu is not a numeric field and will not be compared
```

Also a couple of suggested edits.

abishekg7 commented:
Modified the two parallel constructs to explicitly specify the level of parallelism. This also seems to lead to a performance increase for the limited area test by about 10% in the parallel clause on L1754 and about 51% on L1767. Also verified that results are still bit reproducible.

Added timers around the exit data clause and made the other suggested formatting changes. 

The differences in the diagnostic fields in the history files are expected per our previous discussion.

mgduda commented on src/core_atmosphere/dynamics/mpas_atm_time_integration.F:
```diff
@@ -1710,59 +1710,91 @@ subroutine atm_compute_moist_coefficients( dims, state, diag, mesh, &
+      moist_end = moist_end_ptr
+
+      MPAS_ACC_TIMER_START('atm_compute_moist_coefficients [ACC_data_xfer]')
+      !$acc enter data copyin(qtot, cqw, cqu, scalars)
```
If I'm not mistaken, we only need to `create`, rather than `copyin`, the variables `qtot`, `cqw`, and `cqu`.

mgduda replied:
The line that you referenced only initializes the garbage cell for `qtot`. Setting aside our recent experience with the garbage cell for `theta_m`, I don't think the specific value in the garbage cell for `qtot` should matter, in which case `create` should be fine in the `!$acc data` directive.

abishekg7 replied:
Got it. I changed it the three variables to `create` now, and the results are still bit identical.

mgduda commented:
@abishekg7 The code changes themselves look good to me now. I'd only request that the commit history be cleaned up, and perhaps just a single commit would suffice.

abishekg7 commented:
Rebased this PR and ran a quick check for bit reproducibility again. thanks for the review!

mgduda commented:
> Rebased this PR and ran a quick check for bit reproducibility again. thanks for the review!

Nice work!

gdicker1 reviewed (approved):
This looks good to me!
